docs/guides/auth.md

```diff
@@ -66,45 +79,75 @@ passed to the factory method.
+well as for applications running in [Google Compute Engine
+(GCE)](https://cloud.google.com/compute/).  In the former case, the service
+account’s private keys are loaded from the file named in the environment
+variable `GOOGLE_APPLICATION_CREDENTIALS`. The keys are used to generate bearer
+tokens that are attached to each outgoing RPC on the corresponding channel.
 
 For applications running in GCE, a default service account and corresponding
-OAuth scopes can be configured during VM setup. At run-time, this credential
+OAuth2 scopes can be configured during VM setup. At run-time, this credential
 handles communication with the authentication systems to obtain OAuth2 access
 tokens and attaches them to each outgoing RPC on the corresponding channel.
-Extending gRPC to support other authentication mechanisms
-The gRPC protocol is designed with a general mechanism for sending metadata
-associated with RPC. Clients can send metadata at the beginning of an RPC and
-servers can send back metadata at the beginning and end of the RPC. This
-provides a natural mechanism to support OAuth2 and other authentication
-mechanisms that need attach bearer tokens to individual request.
 
-In the simplest case, there is a single line of code required on the client
-to add a specific token as metadata to an RPC and a corresponding access on
-the server to retrieve this piece of metadata. The generation of the token
-on the client side and its verification at the server can be done separately.
 
-A deeper integration can be achieved by plugging in a gRPC credentials implementation for any custom authentication mechanism that needs to attach per-request tokens. gRPC internals also allow switching out SSL/TLS with other encryption mechanisms.
+### Extending gRPC to support other authentication mechanisms
+
+The Credentials plugin API allows developers to plug in their own type of
+credentials.
+
+- The `MetadataCredentialsPlugin` abstract class contains the pure virtual
+  `GetMetadata` method that needs to be implemented by a sub-class created by
+  the developer.
+- The `MetadataCredentialsFromPlugin` function which creates a `CallCredentials`
+  from the `MetadataCredentialsPlugin`.
+
+Here is example of a simple credentials plugin which sets an authentication
+ticket in a custom header.
+
+```cpp
+class MyCustomAuthenticator : public grpc::MetadataCredentialsPlugin {
+ public:
+  MyCustomAuthenticator(const grpc::string& ticket) : ticket_(ticket) {}
+
+  Status GetMetadata(
+      grpc::string_ref service_url, grpc::string_ref method_name,
+      const AuthContext& channel_auth_context,
+      std::multimap<grpc::string, grpc::string>* metadata) override {
+    metadata->insert(std::make_pair("x-custom-auth-ticket", ticket_));
+    return Status::OK;
+  }
+
+ private:
+  grpc::string ticket_;
+};
+
+auto call_creds =
+    MetadataCredentialsFromPlugin(std::unique_ptr<MetadataCredentialsPlugin>(
```

Comment:
Some items do not have `grpc::`, such as Status, AuthContext, MetadataCredentialsPlugin.